File: catalog/backends/python-django/patterns/n_plus_one_route.py
import psycopg2
import os
from django.http import JsonResponse

# Pattern: DBM N+1 — one query per parent row
# Adapt: replace table/column names with domain entities
DB_URL = os.environ.get("DATABASE_URL", "postgresql://demo:demo@postgresql:5432/demo")


def items_with_details(request):
    conn = psycopg2.connect(DB_URL)
    cur = conn.cursor()
    cur.execute("SELECT * FROM parents")
    parents = cur.fetchall()
    results = []
    for parent in parents:
        cur.execute("SELECT * FROM children WHERE parent_id = %s", (parent[0],))
        results.append({"parent": parent, "children": cur.fetchall()})
    cur.close()
    conn.close()
    return JsonResponse(results, safe=False)

File: catalog/backends/node-express/patterns/n_plus_one_route.js
const { Pool } = require('pg');

// Pattern: DBM N+1 — one query per parent row
// Adapt: replace table/column names with domain entities
const pool = new Pool({
  connectionString: process.env.DATABASE_URL || 'postgresql://demo:demo@postgresql:5432/demo',
});

async function itemsWithDetails(req, res) {
  const { rows: parents } = await pool.query('SELECT * FROM parents');
  const results = [];
  for (const parent of parents) {
    const { rows: children } = await pool.query(
      'SELECT * FROM children WHERE parent_id = $1', [parent.id]
    );
    results.push({ parent, children });
  }
  res.json(results);
}

module.exports = { itemsWithDetails };
